Update CSI attacher to fetch VolumeAttachments from lister

chrishenzie opened the pull request:
**What type of PR is this?**
/kind cleanup
/kind feature

**What this PR does / why we need it**:
This PR changes how the status of VolumeAttachments are checked by the CSI plugin's attacher that runs in the AttachDetachController.

When an attachment is started using the CSI plugin in the AttachDetachController, we poll the VolumeAttachment lister for the updated status. This prevents excess watches from being created via the API (which is the normal behavior), which eats up threads on the API server.

A previous version of this PR attempted to follow an event-based approach by dynamically "hooking into" a SharedInformer, however this introduced a lot of complexity that we were not comfortable maintaining.

The CSI plugin's attacher that runs in Kubelet still uses the event-based approach (API watches) and is not changed, because the Kubelet does not have access to a VolumeAttachment lister, and because using API watches is still more performant than polling via API get calls.

**Which issue(s) this PR fixes**:
Fixes #64429

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig storage

@msau42 @yuga711 @jingxu97

chrishenzie committed:
Interfaces for fake volume host types

chrishenzie committed:
Extend test CSI plugin constructor to support multiple volume host types

chrishenzie commented on pkg/volume/csi/csi_plugin_test.go:
```diff
@@ -41,8 +41,26 @@ import (
 
-// create a plugin mgr to load plugins and setup a fake client
+const (
+	volumeHostType int = iota
```
I'm unsure if this is a great approach. An alternative is passing a FakeVolumeHost constructor as an argument (like `NewFakeVolumeHostWithCSINodeName`, `NewFakeKubeletVolumeHostWithCSINodeName`, etc.)

chrishenzie commented on pkg/volume/csi/shared_informer_hook.go:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	if ch, ok := s.objectNamesToWatch[metaObj.GetName()]; ok {
+		ch <- watch.Event{
```
I was unsure how to handle this.
- If the channel is unbuffered, then this event handler blocks until the channel is read
- If I write in a goroutine, then we risk leaking goroutines that are blocked until the channel is read
- If any of these buffered channels fill, then this event handler blocks everyone until it's read from

Any suggestions are appreciated.

chrishenzie commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -139,6 +146,11 @@ func (c *csiAttacher) WaitForAttach(spec *volume.Spec, _ string, pod *v1.Pod, ti
 
 	attachID := getAttachmentName(source.VolumeHandle, source.Driver, string(c.plugin.host.GetNodeName()))
 
+	if c.plugin.volumeAttachmentInformerHook != nil {
```
I didn't leave a comment here because this method watches for an attach that was already created by the caller.

chrishenzie commented on pkg/volume/csi/csi_attacher_test.go:
```diff
@@ -242,20 +240,18 @@ func TestAttacherAttach(t *testing.T) {
 
 			if tc.expectVolumeAttachment {
-				<-attachmentWatchCreated
+				time.Sleep(testWatchEventDelay)
```
The purpose of this delay is because we can't use the `attachmentWatchCreated` channel when working with a SharedInformer. The prepended watch reactor that would usually notify a watch has been created via the API never fires with a SharedInformer.

There are a few approaches (others are appreciated):

- Add a channel field to the CSI attacher that is written to every time a watch gets created (or fetched from the shared informer hook), just used in tests
  - Pros: Minimal code changes to set this up (as far as I can tell)
  - Cons: Requires adding fields to the CSI attacher just for tests, seems hacky
- Create a custom SharedInformerHook that uses a similar channel-closing mechanism when `GetWatch()` is called
  - Pros: Don't need to modify the CSI attacher
  - Cons: Requires extending the SharedInformerHook, which would require dependency injecting the SharedInformerHook in the CSI plugin
- This approach
  - Pros: Very easy to set up and adjust
  - Cons: Could cause slow or flaky tests

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -106,6 +107,12 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
+		defer c.plugin.volumeAttachmentInformerHook.DeleteWatch(attachID)
+	}
+
 	_, err = c.k8s.StorageV1().VolumeAttachments().Create(context.TODO(), attachment, metav1.CreateOptions{})
```
I was thinking with sharedinformer, create/delete/update/list can all be used with informer instead of directly updating api server?

chrishenzie replied:
I don't think there's a way to make those API calls directly through the SharedInformer.

jingxu97 replied:
sorry my question is not correct. based on https://github.com/kubernetes/kubernetes/issues/64429, before creating/deleting, could use informer (lister) to check whether object exists or not to save some unnecessary calls.

chrishenzie replied:
Oh I see what you mean. I had previously tried doing this in this PR but it introduced some complexity that I wasn't sure how to simplify.

I have some remnants of this approach [in this branch](https://github.com/chrishenzie/kubernetes/compare/volume-attachment-informer...volume-attachment-informer-tip). In order to make this work, we need a way of ensuring an object created through the API is available through the lister. One way we could do this is by prepending a reactor that updates the informer immediately, but we need to plumb that reactor function through to the CSI test plugin constructor (see the `updateInformerReactionFunc` type and changes to `newTestPluginWithVolumeHost()`). I can play around with this again today and see if I can get it working.

chrishenzie commented on pkg/volume/csi/BUILD:
```diff
@@ -12,6 +12,7 @@ go_library(
         "csi_plugin.go",
         "csi_util.go",
         "expander.go",
+        "shared_informer_hook.go",
```
I don't think this is the right place for this type. Should this be tackled separately (is this appropriate for sig apimachinery)?

chrishenzie commented on pkg/volume/csi/csi_attacher_test.go:
```diff
@@ -1630,34 +1613,6 @@ func TestAttacherUnmountDevice(t *testing.T) {
 
-// notifyWatchCreated returns a watch reaction func that closes the supplied channel once a watch is created.
-// This is useful for ensuring a watch has been created prior to injecting events.
-func notifyWatchCreated(watchCreated chan core.Action, watcher *watch.RaceFreeFakeWatcher, targetInvocation int) core.WatchReactionFunc {
```
Since this was introduced earlier and now is removed, I'll see about rebasing this to remove it entirely.

caesarxuchao commented on pkg/volume/csi/shared_informer_hook.go:
```diff
@@ -0,0 +1,112 @@
+	"k8s.io/client-go/tools/cache"
+)
+
+// SharedInformerHook provides a way to dynamically watch objects via a SharedInformer.
```
I'm not sure if it's necessary to add `SharedInformerHook`.
You can create informer to watch object of specific name by use `tweakListOptions` to add [fieldSelector](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L339). IIRC, the fieldSelector for object name is `metadata.name`.

chrishenzie replied:
@caesarxuchao Can I dynamically modify this list of options while the SharedInformer is already running?

I want to allow the CSI plugin's SharedInformer to run, and while it's running I want to dynamically start and stop watching for many objects by name. I need a way of taking the events for a specific tracked object and funneling it to the right channel for that one object. I'm unsure if `tweakListOptions` would enable me to produce this kind of behavior.

caesarxuchao replied:
You are right, that's not supported by the shared Informer. Thanks for the explanation!

yuga711 commented on pkg/volume/csi/shared_informer_hook.go:
```diff
@@ -0,0 +1,112 @@
+func (s *SharedInformerHook) CreateWatch(name string, bufferSize int) <-chan watch.Event {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
```
Don't we need check if an entry for this attachID already exists?

chrishenzie replied:
Good catch, thank you.

chrishenzie commented:
/retest

msau42 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -440,13 +464,22 @@ func (c *csiAttacher) waitForVolumeAttachDetachStatus(attach *storage.VolumeAtta
+	var ch <-chan watch.Event
+	if c.plugin.volumeAttachmentInformerHook != nil {
+		ch = c.plugin.volumeAttachmentInformerHook.GetWatch(attachID)
+	} else {
```
I think the design and logic could be simpler if we didn't have to support kubelet attach/detach mode.

Unfortunately we don't test it so it's not clear if csi attach is even working in kubelet mode attach or not. I think it would fail mainly when trying to Create/Delete VolumeAttachment objects. I expect the node authorizer would reject it if it was coming from kubelet on a node. Can you try an experiment by enabling kubelet Attach/Detach to see if we do hit these errors? https://github.com/kubernetes/kubernetes/blob/87984d84d18c7816835e5efedebb17ed13aa2509/cmd/kubelet/app/options/options.go#L512

cc @jsafrane @saad-ali @gnufied @jingxu97  for thoughts.

chrishenzie commented:
For context, reviewers and I synced offline to discuss what was going on in this PR. Key takeaways below:

- We considered a polling model (polling the lister) instead of this event-based approach
  - Pros: Simpler
  - Cons: Could cause performance regressions
- There are concerns about potentially leaking events if we use buffered channels in the SharedInformerHook
  - The workaround for this is to create channels of size 1 and use the event as a "notification" to go check the `volumeAttachmentLister` for the latest state of the object
    - I've added a comment with the implementation to explain what's going on
- It's better to keep the Kubelet and AttachDetachController specific implementations separate to keep things simpler (plus the implementations have diverged enough)

chrishenzie commented:
/retest

chrishenzie commented:
/retest

chrishenzie commented:
/retest

It seems like this is a flaky test. I encountered failures locally running the master revision of the e2e test.

chrishenzie commented:
/retest

chrishenzie commented:
I think the cause of the flaky test is because the `*WithInformer()` implementation does not check the current state of the object before reading the watch channel. I saw attach timeouts when testing locally, and I think it is because we may have missed the event between subsequent runs of `Attach()`, but this implementation assumes another event would be there (but it wouldn't if the object was already created and updated).

chrishenzie commented:
/retest

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -139,31 +153,118 @@ func (c *csiAttacher) WaitForAttach(spec *volume.Spec, _ string, pod *v1.Pod, ti
+		// Ch is a buffered channel that receives events from the SharedInformerHook. Since there is the
+		// possibility of events being dropped, we ignore the received event and treat it as a notification
+		// to check the lister for the latest status.
+		case _, ok := <-ch:
```
so you are still using event based monitor? The order option is periodically get from lister so that you need to create watch?

chrishenzie replied:
I wanted to try the event-based monitor approach first because there were concerns with performance regressions if we went with a polling model (assuming that meant we would block for some time and repeatedly poll the lister until it was available).

With that being said, the way attach currently works requires the VolumeAttachment object to exist in the lister prior to watching it for an "updated" event. Given this design, what most often happens is:
- `Attach()` runs and creates a VolumeAttachment
- `waitForVolumeAttachmentInternal()` runs, but does not find the VolumeAttachment in the lister (because it was just created, not enough time has passed yet), it returns an error (so no watch gets created)
- `Attach()` runs again and creates a VolumeAttachment and fails (because it exists)
- `waitForVolumeAttachmentInternal()` runs, this time it finds the VolumeAttachment in the lister
- `waitForVolumeAttachDetachStatus()` runs, it sees the VolumeAttachment succeeded and returns (without ever needing to create a watch)

With the way `Attach()` is repeatedly called by the AttachDetachController, we do typically end up repeatedly polling for the VolumeAttachment. If we got rid of the watch logic entirely and just let the ADController repeatedly try until it succeeds we could have more async behavior.

Was what I described above in line with your thoughts on a polling solution?

jingxu97 replied:
Thanks a lot for the detailed explanation. For step 2 and 3, how long it typically take? Another word, how long it takes for the VA object get into the lister?

chrishenzie replied:
Running one of the e2e tests against a kind cluster, I observed it took 55552 nanoseconds ~= 55 microseconds (in the case that a VolumeAttachment was immediately fetched from the lister after creation).

Occasionally the object is not immediately available, in which case we have to wait 100ms for the reconciler to run again.

jingxu97 replied:
got it, thanks!

I think it is find to let it retry here.

jingxu97 replied:
I am trying to think it again about periodic pull approach. Since it is pulling from Lister, what performance regressions you are concern? In think overhead should be small. The time period to pull is set to every say a few second seems fine to me.  
The watch approach is nice too, just in case some race condition, the watch could not get the event because it was deleted just before you start to watch it, then you have to wait until watch timeout which might cause a much longer time.

chrishenzie replied:
I ran the volume stress test a couple times, and observed ~81% of the time we verify attachment during the watch and not via the get from the lister, however the test fails saying not all volumes are bound in time. Comparing this against the base revision now.

I thought there may be performance regressions in CPU because the reconciler would have to call Attach more than once for an attachment to complete. I also thought because this might impact how long it takes for volumes to be attached, however since the reconciler runs 10 times per second I wonder if this impact would even be noticeable.

chrishenzie replied:
@jingxu97 @msau42 

Below is an example of a more asynchronous version of `Attach()` that is designed to be called repeatedly by the reconciler until it succeeds. If we go this route it should cut down on complexity and lines of code. Thoughts on this approach instead?

https://github.com/chrishenzie/kubernetes/blob/volume-attachment-informer-retry/pkg/volume/csi/csi_attacher.go#L66-L142

msau42 replied:
It certainly makes simplifies things, with the main downside that we will retry Attach() multiple times. Do you have a spreadsheet you can share that has the perf comparison?

@jsafrane do you remember why we originally implemented a Watch for the Attach operation? PV controller and resize controller are async operations for CSI as well. It does mess up our metrics a little bit, but they're already broken for provision and resize.

jsafrane replied:
Watch was IMO used only to speed things up and simulate existing behavior of in-tree attach. Volume attachment is part of pod startup and it might be watched by scale team. Will the reconciler add exponential backoff between each Attach call? It grows quickly and it could affect attach speed.

chrishenzie replied:
Yes the reconciler does exponentially back off. When I ran the stress test I saw it wait 500ms between tries.

msau42 replied:
Was this with the mock driver? Could you try injecting a sleep of various magnitudes to the mock driver's attach call? (maybe 1s, 10, 30). I believe with the operation_executor, there is an additional backoff that could get bigger on the order of a minute or more.

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -106,22 +107,33 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
+			return "", errors.New(log("failed to get volume attachment from lister: %v", err))
 		}
-		alreadyExist = true
+		alreadyExist = false
```
just a nit: alreadyExist seems not very necessary here.

If err != nil || va == nil {
    return with error
}
// continue since err== nil and va != nil

jingxu97 replied:
do you want to simplify the logic here if possible?

chrishenzie replied:
Yes, I will get around to this. I'm currently looking at running a little test to see at what point do we successfully verify an attachment has succeeded, because we either successfully verify this at the moment we fetch it from the lister, otherwise we verify it by waiting for an event. If it seems like the majority of the time we successfully verify it the moment we fetch from the lister, then it may be worth removing the watch logic entirely.

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -106,22 +107,33 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
-	_, err = c.k8s.StorageV1().VolumeAttachments().Create(context.TODO(), attachment, metav1.CreateOptions{})
-	alreadyExist := false
+	// Begin watching for this volume attachment prior to creating it to avoid race conditions.
+	if _, err = c.plugin.volumeAttachmentInformerHook.CreateWatch(attachID, watchChannelBufferSize); err != nil {
```
I am not very sure about watch design.

Previously, a watch is created for every unique volumeId. For attach/detach operation, we do not allow concurrent operations. So for each unique volumeID, only one attach or detach operation is allowed. So the total number of watch is equal to the total number of unique volumes. 

For your new design, the number of watch will still be the total number of unique volumes?

chrishenzie replied:
Yes. If we're already watching for a particular VolumeAttachment then this `CreateWatch()` call will fail.

jingxu97 replied:
then what benefit we can get from this?

chrishenzie replied:
To clarify, there will be one API watch for VolumeAttachments, which is the one used by the SharedInformer. We will be "watching for" numerous VolumeAttachments by dynamically hooking into this SharedInformer, but under the hood this is not creating more API watches.

jingxu97 replied:
oh, I think I got it now. Instead of calling API server to create each watch, you use sharedInformer to create those watch.

chrishenzie replied:
The "watch" that we create via this SharedInformerHook is not the same as the "watch" that we create via the API. The "watch" that we create via the SharedInformerHook is just adding an entry to a map of volume attachment IDs that we will keep track of.

chrishenzie commented:
/retest

chrishenzie committed:
Poll for VolumeAttachments in CSI attacher

The CSI attacher that runs inside of the AttachDetachController has
access to a VolumeAttachment lister. By polling this lister for the
status of VolumeAttachments, we can save threads on the API server by
not using watches.

chrishenzie commented:
This latest version of the PR polls the lister for the status of VolumeAttachments (instead of watching for changes). Doing this seems to have reduced a lot of the complexity involved in "hooking" into the SharedInformer to watch for certain objects. 

- The `Attach()` operation was refactored to check if a VolumeAttachment exists in the lister first so we save an API call (in the case that the object already exists, and would return an error).
- There are separate methods for polling ("waiting") for the status of attachments and detachments, in order to simplify things.
- This approach has removed the need to prepend reactors (in tests) to update the lister, since the "wait" implementation will poll until the object is available. The nice part about this is that we're now testing that objects created through the API propagate to the lister.

Let me know if you have any questions.

yuga711 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -166,6 +168,46 @@ func (c *csiAttacher) waitForVolumeAttachmentInternal(volumeHandle, attachID str
+			volumeAttachment, err := c.plugin.volumeAttachmentLister.Get(attachID)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					// Ignore "not found" errors in case the VolumeAttachment was just created and hasn't yet made it into the lister.
```
Can we add a log message? I think should be very rare..

chrishenzie replied:
If we move this check to the top of the loop (outside of the ticker), then this log message seems very likely, because we would start checking the lister immediately after creating the object in the API, and the staleness latency of the lister could be hundreds of milliseconds or more depending on API server load.

yuga711 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -166,6 +168,46 @@ func (c *csiAttacher) waitForVolumeAttachmentInternal(volumeHandle, attachID str
+	for {
+		select {
+		case <-ticker.C:
+			volumeAttachment, err := c.plugin.volumeAttachmentLister.Get(attachID)
```
You may want to move this before select{} so that super fast attachments doesnt need to wait for first timer tick :)

chrishenzie replied:
Thanks for the suggestion, I like this

chrishenzie replied:
The other nice thing about this approach is it removed the need to use labels. I had to use them previous to break out of the loop because `go vet` complained about unreachable code.

yuga711 commented:
You may want to update the PR description to match the current implementation.

chrishenzie commented:
Updated, I tried logging in the operation executor how long the attachment takes since the end-to-end timing seems to include more than what we're trying to measure. [See here for the results](https://gist.github.com/chrishenzie/237e9b06fffdbd3bbfd07c6dee121830).

chrishenzie commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -76,52 +77,53 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
-		vaSrc = storage.VolumeAttachmentSource{
-			PersistentVolumeName: &pvName,
-		}
+	attachment, err := c.plugin.volumeAttachmentLister.Get(attachID)
```
I checked with someone on API machinery who suggested the staleness of data is usually on the order of hundreds of milliseconds, but in worst case scenarios it could be over a second.

In what scenario do you think this staleness could cause a problem?

yuga711 replied:
I do not know much about API machinery. Please ignore if you have already checked with an expert in this area. My scenario is:  Let’s say vol1 is attached to node1 and is then detached and is being attached now. Is it possible for VALister to return success because of stale cache?

chrishenzie replied:
Yes it would be possible for the lister to return stale data which could cause it to think the VolumeAttachment already exists when it was just deleted. In this scenario the request would time out and then be retried next reconciler loop.

In the other race condition scenario where the VolumeAttachment doesn't show up in the lister (yet) but exists in the API, we ignore any "AlreadyExists" errors returned by the API and proceed with polling for the object until it's status shows a successful attachment.

jingxu97 replied:
if you are concern the case that VA still exist, but volume is already detached, I think the following verifyVolumeAttachment will check VA deletionTimeStamp is already set, and will return with error

chrishenzie commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -76,52 +77,53 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
+			},
+		}
+
+		_, err = c.k8s.StorageV1().VolumeAttachments().Create(context.TODO(), attachment, metav1.CreateOptions{})
```
From my understanding, the reconciler can only process one attachment per attachment ID at a time, so in order for that to happen:

- `Attach()` gets called once, creates attachment
- An error gets thrown while waiting for the attachment, which causes the reconciler to try the attachment again
- `Attach()` gets called again (faster than the "staleness latency" of the lister) and so it tries to create the attachment
- Attachment creation fails and the reconciler tries again until the object is present in the lister

If checking the lister before creating the object adds too much complexity in terms of thinking about possible race conditions, then we can go back to the old behavior of trying to create the object, the only tradeoff is extra create calls to the API.

jingxu97 replied:
after Create call, maybe use the same logic as before, if error type is alreadyExist, it can continue to waitForVolumeAttach instead of let reconciler to retry (this might cause a little bit delay due to exponential backoff)?

yuga711 replied:
Looks like AlreadyExist error is now being handled below.

chrishenzie commented:
@yuga711 I addressed the comments above, thank you for the review. The only thing untouched is whether or not to check the lister first in `Attach()`, or to rely on the previous behavior of checking the API directly. I'm comfortable with making that change, just let me know your thoughts. I also updated the description, title, and the gist above (containing latency metrics) to reflect performance with these new changes.

chrishenzie commented:
/retest

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -166,6 +168,43 @@ func (c *csiAttacher) waitForVolumeAttachmentInternal(volumeHandle, attachID str
+	for {
+		volumeAttachment, err := c.plugin.volumeAttachmentLister.Get(attachID)
+		if err != nil {
+			// Ignore "not found" errors in case the VolumeAttachment was just created and hasn't yet made it into the lister.
```
maybe first wait for pollInterval before get from lister to help avoid always retrying?

chrishenzie replied:
This was the original approach. [See this comment thread](https://github.com/kubernetes/kubernetes/pull/97834#discussion_r569994822) for context on why I changed it. I can always move it back if you feel this is a waste of a timer tick.

jingxu97 replied:
I remember you did some experiments about this. One tick is 500 ms. You can decide which way to go based on your data?

chrishenzie replied:
I tuned the ticker to be a bit more aggressive, this time 100ms. [These are the timings](https://gist.github.com/chrishenzie/2ed57c1a6bf45770c1bdffc93ee48036), compared with the [old timings when it was 500ms](https://gist.github.com/chrishenzie/237e9b06fffdbd3bbfd07c6dee121830). The speedup is somewhat negligible, around half a second to a second faster depending on test noise.

chrishenzie replied:
See the comment thread below. It seems like having an aggressive frequency (like 100ms) doesn't make sense since we're waiting on the result of a storage provider's API call which may take >10s. This should be tuned to something less frequent, like maybe 1 second.

jingxu97 replied:
what I mean is explicitly wait a little bit before getting to get VA from lister. I remember you mentioned that typically for the first time, this function will create VA, and then try to get VA from lister, which will fail mostly because VA has not yet in lister. Will it bring some benefit to wait a second so that lister will have the VA already when we check it here?

jingxu97 replied:
also I agree 100ms is too aggressive, starting from 1 sec seems reasonable.

chrishenzie replied:
Ok, I've bumped it to 1 second

msau42 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -166,6 +171,43 @@ func (c *csiAttacher) waitForVolumeAttachmentInternal(volumeHandle, attachID str
 	return attach.Name, nil
 }
 
+func (c *csiAttacher) waitForVolumeAttachmentWithLister(volumeHandle, attachID string, pollInterval, timeout time.Duration) error {
```
What is the timeout? If it's too long, then we may end up excessively spinning too much if the apiserver is slow/having problems, especially with multiple attachements happening in parallel. This is the main reason why we do exponential backoff.

In general we've tried to move away from having retry loops within retry loops because it makes it difficult to reason about the behavior.

msau42 replied:
In your benchmarking test, it would be good to also inject a sleep into the mock driver to simulate a more realistic volume plugin with a longer attach time.

yuga711 replied:
If needed, we could probably double the pollInterval after every tick. If we do that we can probably start with small interval and exponentially increase after every tick.

chrishenzie replied:
Discussed this offline. The reasons for keeping this synchronous for now was in order to stay consistent with the attachment behavior used by our other (in-tree) plugins, and also because making this asynchronous adds a lot of complexity to our tests. That work is probably best tackled in a separate PR.

Please see the comment below on thoughts around using exponential backoff vs. fixed poll intervals.

chrishenzie replied:
Also for context, I tried adding a `time.Sleep(40 * time.Second)` in the mock CSI driver `ControllerPublishVolume()` call, however I encountered other issues with the mock CSI driver that prevented this experiment from working. This requires more digging to understand what's going on.

chrishenzie commented:
/retest

chrishenzie commented:
/retest

These tests are passing for me locally on both master and my branch.

chrishenzie commented:
I experimented with a solution that uses exponential backoff (instead of a fixed poll interval), and collected [these attachment timings](https://gist.github.com/chrishenzie/4d5a08cdcfd0400ac038b5998c9e5da6). This is with a mock driver with `ControllerPublishVolume()` returning immediately. My backoff was tuned as follows:
```go
// Configured to try at 15ms, 90ms, 540ms, 3.24s, 19.44s, 116.64s.
attachBackoff := wait.Backoff{
        // The backoff cap will be hit before we reach this many iterations.
        Steps:    7,
        Duration: 15 * time.Millisecond,
        Factor:   6.0,
        Jitter:   0.1,
        Cap:      timeout,
}
```

With this set of backoff parameters, it took ~2.5x longer for all attachments to complete. We could tune this further, however I'm unsure if exponential backoff is the right approach.

It seems like we want our time-to-attach to resemble the latency of the storage provider's API as closely as possible, and with exponential backoff we run the risk of introducing additional latency.

Assuming most storage provider APIs will take potentially 10s of seconds to return for `ControllerPublishVolume()`, what if we change this to use something like `wait.PollImmediate()` and change the poll interval to something less aggressive, like once a second. This mimics [the approach used in the PD CSI driver](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/95ffc203e9aabb8d11f3d7d1572d82ea5e5d3b71/pkg/gce-cloud-provider/compute/gce-compute.go#L900) when querying for the status of a snapshot.

chrishenzie commented:
/retest

msau42 commented:
> // Configured to try at 15ms, 90ms, 540ms, 3.24s, 19.44s, 116.64s.

I think this is backing off too aggressively. Can we have max backoff period to be around 10-15 seconds? I also think we could start the first retry period at something much higher like 100ms+

chrishenzie commented:
@msau42 Do you mean reducing the CSI timeout to 10-15 seconds? This will increase the likelihood that an `Attach()` call will need to be performed several times in order to succeed (for slow storage provider APIs).

chrishenzie commented:
For tuning this further, [this is a small program](https://gist.github.com/chrishenzie/604d6f9e744cdac7c6bd353336896a92) that prints out the elapsed time since the start, and the elapsed time between loop iterations. [Here are the timings](https://gist.github.com/chrishenzie/479462b0b9d8cf89d53e081b3b22513e) for the most recent run.

chrishenzie commented:
/retest

chrishenzie commented:
This latest revision uses exponential backoff with a maximum backoff of 5 seconds.

Below are attachment timings collected for 0 second, 10 second, and 30 second waits in the mock CSI driver during `ControllerPublishVolume()`. It includes timings comparing the previous 15 second `Cap` exponential backoff with this 5 second max exponential backoff. Both are using the same initial duration of 1 second, growth factor of 1.15, and jitter of 0.1.

- [0 second wait]
- [10 second wait]
- [30 second wait]

[0 second wait]: https://gist.github.com/chrishenzie/1218a57e5b5742b47b50476b89432b3b
[10 second wait]: https://gist.github.com/chrishenzie/647a45992ecc5b6e28694eee917cf9ea
[30 second wait]: https://gist.github.com/chrishenzie/62c44eb6b13c519243db449603d2af9c

chrishenzie commented:
/retest

chrishenzie commented:
/retest

msau42 commented:
@chrishenzie can you also link your most recent spreadsheet here?

chrishenzie commented:
@msau42 [Here's the spreadsheet with the attachment timings](https://github.com/kubernetes/kubernetes/pull/97834). Included are columns for timings with GCE PD. Coincidentally this branch ended up performing faster, however I'm re-running the test a few more times to get a better comparison of averages.

chrishenzie commented:
I ran the test three more times for both the base revision + this branch. For 30 pods (90 samples total), the base revision took on average 26,338.46ms, and this branch took on average 25,773.48ms, which is about a ~2% speedup.

msau42 commented:
I was surprised the PD driver results were very different than the mock driver with similar delay, but the mock driver only works on a single node so maybe that is coming into play. I think the results are looking good, they are within < 5% noise or better. Thanks for being patient and running all the various benchmarking experiments!

/lgtm
/approve

yarinm commented:
@chrishenzie
We recently started experiencing delays in volume attachments for pods, when we describe the pods we see them in a loop of failures regarding:

```
 Warning  FailedAttachVolume  61s (x9 over 5m11s)      attachdetach-controller                 AttachVolume.Attach failed for volume "6401e5d7-ec8e-4aff-98f0-e9694436e8b1-ab1498ab42d22c11bcbe7fd5d5" : volume attachment watch channel had been closed 
``` 

There are no logs in external-attacher or our CSI-driver for any work regarding the PV and the only logs about the PV in kube-controller-manager are with the same watch channel closed error and nothing else.

Do you think your PR could potentially solve our issue? If so can this be backported to 1.20?

I'm using AKS (1.18) with external-attacher (3.1.0) we have around 500-800 PVs that are needed to be handled by CSI

chrishenzie commented:
@yarinm
Do your API server logs tell anything? I would expect if the channel is unexpectedly closing that it's because the API server is closing the connection prematurely.

I think what's likely happening is the attach-detach-controller is opening 500-800 watch connections for each PV and trying to listen for changes, but the API server gets overloaded and starts dropping connections, resulting in the behavior you're seeing.

If this is the case, then I think this change would fix the issue, and we should prioritize patching it in as a bugfix in older minor versions.

chrishenzie commented:
When these attachments are hanging, do you see any corresponding `VolumeAttachment` objects?
```
kubectl get VolumeAttachment
```

If you do not observe any `VolumeAttachments` in the API then I would understand why the external-attacher doesn't contain any logs for this, however if `VolumeAttachments` weren't being created then I would expect to see error logs in the kube-controller-manager telling that the object could not be created.

Also, do these attachments eventually succeed, or are they indefinitely stuck like this?

yarinm commented:
They are eventually processed, it varies from several minutes to couple of hours depending on the load. Regarding the attachments I'll have to check when I'll have some time and see what is the exact behaviour.
